climada/hazard/centroids/centr.py

```diff
@@ -555,60 +555,47 @@ def union(self, *others):
-        is_on_land = np.any([len(cent.on_land) > 0 for cent in cent_list]) \
-            and not np.all([len(cent.on_land) > 0 for cent in cent_list])
-        is_region_id = np.any([cent.region_id.any() for cent in cent_list]) \
-            and not np.all([cent.region_id.any() for cent in cent_list])
-        is_elevation = np.any([cent.elevation.any() for cent in cent_list]) \
-            and not np.all([cent.elevation.any() for cent in cent_list])
+        # restrict to non-empty centroids
+        cent_list = [c for c in (self,) + others if c.size > 0 or c.meta]
+        if len(cent_list) == 0 or len(others) == 0:
+            return copy.deepcopy(self)
 
+        # check if all centroids agree
+        if all([cent_list[0].equal(cent) for cent in cent_list[1:]]):
+            return copy.deepcopy(cent_list[0])
+
+        # convert all raster centroids to point centroids
         for cent in cent_list:
             if cent.meta and not cent.lat.any():
                 cent.set_meta_to_lat_lon()
+
+        # make sure that all Centroids have the same CRS
+        for cent in cent_list:
             if cent.crs is None:
                 cent.geometry = cent.geometry.set_crs(DEF_CRS)
-            cent.set_geometry_points()
             if not u_coord.equal_crs(cent.crs, cent_list[0].crs):
-                raise ValueError(f'Different CRS {cent.crs}!='
-                                 f'{cent_list[0].crs} are not accepted.')
-            if is_area_pixel and not cent.area_pixel.any():
-                cent.set_area_pixel()
-            if is_dist_coast and not cent.dist_coast.any():
-                cent.set_dist_coast()
-            if is_on_land and not len(cent.on_land) > 0:
-                cent.set_on_land()
-            if is_region_id and not cent.region_id.any():
-                cent.set_region_id()
-            if is_elevation and not cent.elevation.any():
-                cent.set_elevation()
-
-        #concatenate attributes
-        centroids= Centroids()
-        for attr_name in vars(self).keys():
+                raise ValueError('In a union, all Centroids need to have the same CRS: '
+                                 f'{cent.crs} != {cent_list[0].crs}')
+
+        # set attributes that are missing in some but defined in others
+        for attr in ["geometry", "area_pixel", "dist_coast", "on_land", "region_id", "elevation"]:
+            if np.any([getattr(cent, attr).size > 0 for cent in cent_list]):
+                for cent in cent_list:
+                    if not getattr(cent, attr).size > 0:
+                        fun_name = f"set_{attr}{'_points' if attr == 'geometry' else ''}"
+                        getattr(Centroids, fun_name)(cent)
+
+        # create new Centroids object and set concatenated attributes
+        centroids = Centroids()
+        centroids.meta = {}
+        for attr_name, attr_val in vars(cent_list[0]).items():
             attr_val_list = [getattr(cent, attr_name) for cent in cent_list]
```

Comment:
Technically, this could be moved in the if clauses such that it is not computed when the attribute is neither a 1D numpy array nor a geoseries.